stdlib/list.ml

```diff
@@ -78,7 +80,7 @@ let rev_init_threshold =
 
 let init len f =
   if len < 0 then invalid_arg "List.init" else
-  if len > rev_init_threshold then rev (init_tailrec_aux [] 0 len f)
+  if len > rev_init_threshold then init_tailrec_aux 0 len f
   else init_aux 0 len f
```

Comment:
Honestly I don't really like this version of the code. It is more complex than before, when TRMC is designed to let us write simpler code. I would want this 10_000 stuff to go away, I think that it would be acceptable to do with just a direct-style TRMC version, and again if you want to be absolutely performance-preserving on all lengths you can do a 2-unfolding and get a much simpler and still faster version.

Reply:
Note: it's not clear to me that TRMC (without unfolding) will be slower than this version on small sizes, because that small-size version is paying the extra cost (compared to List.map) of incrementing the depth counter and checking.

Reply:
In both the case of `init` and `of_seq`, the (unfolded) TRMC version "loses" against the standard version in the non-tail-recursive range (and as expected, wins big beyond it). The difference is greater in the case of `init` (I guess because `of_seq` is doing more work due to the need to evaluate the seq). Perhaps we don't care?

Results at https://github.com/nojb/list_trmc/commit/4db39bc965cbeccd56c693b550049042ff7ce90a

Reply:
The following tweak on your list_trmc version seems to consistently beat the stdlib according to your bench:

```ocaml
let[@tail_mod_cons] rec init i last f =
  if i > last then []
  else if i = last then [f i]
  else f i :: f (i + 1) :: init (i + 2) last f

let init len f =
  if len < 0 then invalid_arg "List.init" else
  init 0 (len - 1) f
```

It's also much simpler / more readable.

Reply:
Indeed, amended.

Reply:
Nitpick: I wish you would have used this proposed version of `List.init` which uses `last` instead of `n` for the variable that is equal to `len - 1` initially, and is thus slightly more readable than your current proposal.

Reply:
Good point, fixed.